zebra-network/src/peer_set/set.rs

```diff
@@ -88,12 +95,14 @@ where
     /// These guards are checked for errors as part of `poll_ready` which lets
     /// the `PeerSet` propagate errors from background tasks back to the user
     guards: futures::stream::FuturesUnordered<JoinHandle<Result<(), BoxedStdError>>>,
+    /// Stream of incoming inventory hashes to
+    inv_stream: broadcast::Receiver<(InventoryHash, SocketAddr)>,
+    inventory_registry: InventoryRegistry,
```

Comment:
How do you feel about combining these? The `poll_ready` function is already pretty large at 50 lines, tho most of it is noise from logging, metrics or error handling.

I was thinking we could put the receiver and the interval inside of the registry and only expose a `synchronize()` function instead of `register` and `prune`, which would still exist but would be private methods on the `InventoryRegistry` type. This `sync` function would then do the logic for draining the receiver, checking the interval, and rotating the storage, basically any logic we want to have happen on every `poll_ready` call.

Seems like isolating all of this logic to 1 file would keep the `poll_ready` function from getting unwieldy.

Reply:
Sure, I think there are two options here:

1. Roll all of the registry bookkeeping and updates into the `InventoryRegistry` type as you describe. I think it would be better to call the method `poll_inventory` rather than `synchronize`, in order to convey that it's a poll method that needs a context parameter.  The only concern I have is whether we'd run into pinning issues or something.

2. Roll all of the registry bookkeeping and updates into a `poll_inventory` method on the `PeerSet`.  This would be similar to the `poll_discover` implementation, in fact the full content of the `poll_ready` implementation of the peer set is much longer than 50 lines already, it's just broken up into sub-methods.

Either way I think that you're right that it would be better to isolate the logic to one method, I'm just not sure whether it will be easier to add to the inventory registry or to the peer set.

Reply:
Sounds good, I think that's slightly cleaner too.